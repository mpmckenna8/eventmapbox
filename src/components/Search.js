import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {connect} from 'react-redux';
import Geocoder from './Geocoder';
import PlaceName from './PlaceName';
import CloseButton from './CloseButton';
import PlaceInfo from './PlaceInfo';
import directionsIcon from '../assets/directions.svg';
import {triggerMapUpdate, setDirectionsLocation, getPlaceInfo, resetStateKeys, setStateValue} from '../actions/index';

class Search extends Component {
  render() {
    let SearchBar;

    if (this.props.searchLocation === null) { // no place was selected yet
      SearchBar = (
        <Geocoder
          onSelect={(data) => this.onSelect(data)}
          searchString={this.props.searchString}
          writeSearch={(value) => {
            this.props.triggerMapUpdate();
            this.props.writeSearch(value);
          }}
          resultsClass={this.styles.results}
          inputClass={this.styles.input}
          focusOnMount={true}
        />
      );

    } else { // There is a selected place
      SearchBar = (
        <div className={this.styles.input + ' flex-parent flex-parent--center-cross flex-parent--center-main'}>
          <div className='w-full w420-mm pr42 txt-truncate'>
            <PlaceName
              location={this.props.searchLocation}
              onClick={() => {
                this.props.writeSearch(this.props.searchLocation.place_name);
                this.props.resetStateKeys(['searchLocation', 'placeInfo']);
              }}
            />
          </div>
          <div
            id='search-directions'
            className={'mr30 cursor-pointer right ' + this.styles.icon}
            onClick={() => this.clickDirections()}
          >
            <img src={directionsIcon} alt='directions'/>
          </div>
        </div>
      );

    }

    return (
      <div className={this.styles.mainDiv}>
        <div className={this.styles.icon}>
          <svg className='icon color-gray'><use xlinkHref='#icon-search'></use></svg>
        </div>
        {SearchBar}
        <CloseButton
          show={(this.props.searchString !== '' || this.props.searchLocation !== null)}
          onClick={() => this.closeSearch()}
        />

        {
          (this.props.searchLocation && this.props.placeInfo)
          ? <PlaceInfo info={this.props.placeInfo} clickDirections={() => this.clickDirections()}/>
          : null
        }
      </div>
    );
  }

  onSelect(data) {
    this.props.writeSearch(data.place_name);
    this.props.setSearchLocation(data);
    this.props.triggerMapUpdate('repan');
    if (data.properties.wikidata) this.props.getPlaceInfo(data.properties.wikidata);
  }

  closeSearch() {
    this.props.resetStateKeys(['searchString', 'searchLocation', 'placeInfo']);
    this.props.triggerMapUpdate();
  }

  clickDirections() {
    this.props.setMode('directions');
    this.props.writeSearch(this.props.searchLocation.place_name);
    this.props.setDirectionsLocation('to', this.props.searchLocation);
  }


  get styles() {
    return {
      mainDiv: 'absolute h42 w-full w420-mm bg-white shadow-darken25 flex-parent flex-parent--row flex-parent--space-between-main searchDiv',
      icon: 'absolute flex-parent flex-parent--center-cross flex-parent--center-main w42 h42',
      input: 'input px42 h42 border--transparent',
      results: 'results bg-white shadow-darken25 mt6 border-darken10'
    };
  }
}

Search.propTypes = {
  getPlaceInfo: PropTypes.func,
  placeInfo: PropTypes.object,
  resetStateKeys: PropTypes.func,
  searchLocation: PropTypes.object,
  searchString: PropTypes.string,
  setDirectionsLocation: PropTypes.func,
  setMode: PropTypes.func,
  setPlaceInfo: PropTypes.func,
  setSearchLocation: PropTypes.func,
  triggerMapUpdate: PropTypes.func,
  writeSearch: PropTypes.func,
};

const mapStateToProps = (state) => {
  return {
    placeInfo: state.app.placeInfo,
    searchLocation: state.app.searchLocation,
    searchString: state.app.searchString,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    getPlaceInfo: (id) => dispatch(getPlaceInfo(id)),
    resetStateKeys: (keys) => dispatch(resetStateKeys(keys)),
    setDirectionsLocation: (kind, location) => dispatch(setDirectionsLocation(kind, location)),
    setMode: (mode) => dispatch(setStateValue('mode', mode)),
    setPlaceInfo: (info) => dispatch(setStateValue('placeInfo', info)),
    setSearchLocation: (location) => dispatch(setStateValue('searchLocation', location)),
    triggerMapUpdate: (repan) => dispatch(triggerMapUpdate(repan)),
    writeSearch: (input) => dispatch(setStateValue('searchString', input)),
  };
};

export {Search};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Search);
