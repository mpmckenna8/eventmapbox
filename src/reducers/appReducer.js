import polyline from '@mapbox/polyline';

import  eventsGeojson from '../assets/protests.js';


console.log('events file says', ((eventsGeojson.features)))

const defaultAppState = {
  // Mapbox Access Token
  mapboxAccessToken: process.env.REACT_APP_MAPBOX_TOKEN,
  // Map
  mapCoords: [-122.4, 37.8, 11],
  mapStyle: 'streets',
  // Map updates
  needMapUpdate: false,
  needMapRepan: false,
  needMapRestyle: false,
  // Mode
  mode: 'events',
  modality: 'car',
  // Search
  searchString: '',
  searchLocation: null,
  placeInfo: null,
  // User
  userLocation: null,
  // Directions
  directionsFromString: '',
  directionsFrom: null,
  directionsToString: '',
  directionsTo: null,
  route: null,
  routeStatus: 'idle',
  lastQueried: 0,
  // Context menu
  contextMenuActive: false,
  contextMenuCoordinates: null,
  contextMenuPosition: null,
  contextMenuPlace: null,
  eventsGeojson: eventsGeojson,
  eventLocation: {}
};

const appReducer = (state = defaultAppState, action) => {
  switch (action.type) {

  case 'SET_STATE_VALUE': {
    const modifiedState = {};
    modifiedState[action.key] = action.value;
    return Object.assign({}, state, modifiedState);
  }

  case 'SET_STATE_VALUES': {
    return Object.assign({}, state, action.modifiedState);
  }

  case 'RESET_STATE_KEYS': {
    const modifiedState = {};
    action.keys.forEach(k => {
      modifiedState[k] = defaultAppState[k];
    });
    return Object.assign({}, state, modifiedState);
  }

  case 'TRIGGER_MAP_UPDATE':
    return Object.assign({}, state, {
      needMapUpdate: true,
      needMapRepan: action.needMapRepan
    });

  case 'SET_USER_LOCATION':
    return Object.assign({}, state, {
      userLocation: {
        'place_name': 'My Location',
        center: action.coordinates,
        geometry: {
          type: 'Point',
          coordinates: action.coordinates
        }
      },
    });

  case 'SET_DIRECTIONS_LOCATION': {
    if (action.kind === 'from') {
      return Object.assign({}, state, {
        directionsFrom: action.location
      });
    } else if (action.kind === 'to') {
      return Object.assign({}, state, {
        directionsTo: action.location
      });
    } else return state;
  }

  case 'SET_ROUTE': {
    if (action.data.routes.length > 0 && state.directionsFrom && state.directionsTo) {
      const route = action.data.routes[0];

      let congestion;
      if (route.legs[0] && route.legs[0].annotation && route.legs[0].annotation.congestion) {
        congestion = route.legs[0].annotation.congestion;
      }

      const line = polyline.toGeoJSON(route.geometry);

      if (!congestion) {
        route.geometry = line;
      } else {
        route.geometry = congestionSegments(line, congestion);
      }

      return Object.assign({}, state, {
        route: route
      });
    } else {
      return Object.assign({}, state, {
        routeStatus: 'error'
      });
    }
  }

  case 'SET_CONTEXT_MENU': {
    return Object.assign({}, state, {
      contextMenuCoordinates: action.coordinates,
      contextMenuPosition: action.position,
      contextMenuPlace: {
        'place_name': '__loading', // will trigger the spinner
        center: action.coordinates,
        geometry: {
          type: 'Point',
          coordinates: action.coordinates
        }
      },
      contextMenuActive: true
    });
  }

  case 'RESET_CONTEXT_MENU': {
    return Object.assign({}, state, {
      contextMenuCoordinates: null,
      contextMenuPosition: null,
      contextMenuPlace: null,
      contextMenuActive: false
    });
  }

  case 'SET_STATE_FROM_URL': {
    return state;
  }

  default:
    return state;
  }
};

function congestionSegments(line, congestionArray) {
  let featureCollection = {
    type: 'FeatureCollection',
    features: []
  };
  const coordinates = line.coordinates;
  for (var i = 0; i < coordinates.length - 1; i++) {
    let segment = {
      type: 'Feature',
      geometry: {
        type: 'LineString',
        coordinates: coordinates.slice(i, i + 2)
      },
      properties: {
        congestion: congestionArray[i]
      }
    };
    featureCollection.features.push(segment);
  }
  return featureCollection;
}

export default appReducer;
export {appReducer, defaultAppState};
