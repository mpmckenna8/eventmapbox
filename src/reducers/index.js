import {appReducer, defaultAppState} from './appReducer';

const reducers = {
  app: appReducer
};

const defaultState = {
  app: defaultAppState
};

export default reducers;

export {defaultState};
