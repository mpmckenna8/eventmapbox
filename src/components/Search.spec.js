import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';
import {defaultState} from '../reducers/index';
import reducers from '../reducers/index';
import {Search} from './Search';
import ConnectedSearch from './Search';
import {placeInfo, searchLocation} from './fixtures/Search.fixtures.js';

it('renders correctly with search location', () => {

  const component = renderer.create(
    <Search
      getPlaceInfo={() => {}}
      placeInfo={placeInfo}
      resetStateKeys={() => {}}
      searchLocation={searchLocation}
      searchString='Paris'
      setDirectionsLocation={() => {}}
      setMode={() => {}}
      setPlaceInfo={() => {}}
      setSearchLocation={() => {}}
      triggerMapUpdate={() => {}}
      writeSearch={() => {}}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly without search location', () => {

  let store = createStore(combineReducers({...reducers}), defaultState);

  const component = renderer.create(
    <Provider store={store}>
      <ConnectedSearch/>
    </Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('behaves correctly on clicking directions', () => {

  const setDirectionsLocation = jest.fn();
  const setMode = jest.fn();

  const search = shallow(
    <Search
      getPlaceInfo={() => {}}
      placeInfo={placeInfo}
      resetStateKeys={() => {}}
      searchLocation={searchLocation}
      searchString='Paris'
      setDirectionsLocation={setDirectionsLocation}
      setMode={setMode}
      setPlaceInfo={() => {}}
      setSearchLocation={() => {}}
      triggerMapUpdate={() => {}}
      writeSearch={() => {}}
    />
  );

  search.find('#search-directions').simulate('click');
  expect(setDirectionsLocation).toBeCalledWith('to', searchLocation);
  expect(setMode).toBeCalledWith('directions');
});
