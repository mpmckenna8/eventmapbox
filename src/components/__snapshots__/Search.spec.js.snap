exports[`test renders correctly with search location 1`] = `
<div
  className="absolute h42 w-full w420-mm bg-white shadow-darken25 flex-parent flex-parent--row flex-parent--space-between-main">
  <div
    className="absolute flex-parent flex-parent--center-cross flex-parent--center-main w42 h42">
    <svg
      className="icon color-gray">
      <use
        xlinkHref="#icon-search" />
    </svg>
  </div>
  <div
    className="input px42 h42 border--transparent flex-parent flex-parent--center-cross flex-parent--center-main">
    <div
      className="w-full w420-mm pr42 txt-truncate">
      <div
        className="txt-truncate w-full"
        onClick={[Function]}>
        <div
          className="inline pr6 color-black">
          Paris
        </div>
        <div
          className="inline txt-s color-darken50">
          France
        </div>
      </div>
    </div>
    <div
      className="mr30 cursor-pointer right absolute flex-parent flex-parent--center-cross flex-parent--center-main w42 h42"
      id="search-directions"
      onClick={[Function]}>
      <img
        alt="directions"
        src="directions.svg" />
    </div>
  </div>
  <div
    className="absolute right flex-parent flex-parent--center-cross flex-parent--center-main w42 h42 cursor-pointer"
    onClick={[Function]}>
    <svg
      className="icon  color-gray">
      <use
        xlinkHref="#icon-close" />
    </svg>
  </div>
  <div
    className="place-info absolute top bg-white w-full w420-mm shadow-darken25 flex-parent flex-parent--column">
    <img
      alt="Capital city of France"
      className="cover w-full h72 h120-mm"
      onError={[Function]}
      src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/7f/A/640px-A" />
    <div
      className="bg-blue flex-parent flex-parent--row">
      <div
        className="p6 flex-child flex-child--grow flex-parent flex-parent--column flex-parent--center-main">
        <div
          className="color-white pl42 pr12 txt-h4 txt-bold">
          Paris
        </div>
        <div
          className="color-lighten75 pl42 pr12 txt-s ">
          Capital city of France
        </div>
      </div>
      <div
        className="bg-white hmin42 wmin42 hmin48-mm wmin48-mm hmax42 wmax42 hmax48-mm wmax48-mm m6 m12-mm round-full shadow-darken10 cursor-pointer flex-parent flex-parent--center-main flex-parent--center-cross"
        onClick={[Function]}>
        <img
          alt="directions"
          src="directions.svg" />
      </div>
    </div>
  </div>
</div>
`;

exports[`test renders correctly without search location 1`] = `
<div
  className="absolute h42 w-full w420-mm bg-white shadow-darken25 flex-parent flex-parent--row flex-parent--space-between-main">
  <div
    className="absolute flex-parent flex-parent--center-cross flex-parent--center-main w42 h42">
    <svg
      className="icon color-gray">
      <use
        xlinkHref="#icon-search" />
    </svg>
  </div>
  <div
    className="w-full">
    <input
      className="input px42 h42 border--transparent"
      onChange={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      placeholder="Search"
      type="text"
      value="" />
  </div>
</div>
`;
